### PR TITLE
Handle CNAME in SoA records

### PR DESCRIPTION
**Pull request by tsellers-r7**
This PR addresses an exception caused when performing a query against a DNS name that doesn't itself have a SoA but instead includes a CNAME of a parent domain that contains the SoA. In other words there are CNAMEs in the results that eventually chain to an MNAME that is also in the results and there is no `Authority` section.

The exception looks like this:
```
Traceback (most recent call last):
  File "/Users/UserName/git/dnsrecon/dnsrecon.py", line 1699, in <module>
    main()
  File "/Users/UserName/git/dnsrecon/dnsrecon.py", line 1530, in main
    std_enum_records = general_enum(res, domain, xfr, goo, bing, spf_enum, do_whois, do_crt, zonewalk)
  File "/Users/UserName/git/dnsrecon/dnsrecon.py", line 943, in general_enum
    found_soa_records = res.get_soa()
  File "/Users/UserName/git/dnsrecon/lib/dnshelper.py", line 217, in get_soa
    name = rdata[0].mname.to_text()
AttributeError: 'CNAME' object has no attribute 'mname'
```

This was previously reported in https://github.com/darkoperator/dnsrecon/issues/87

The PR differs from the proposed fix in that Issue in that this PR continues processing and extracts the correct SoA instead of returning an empty list.

The logic behind the changes

- Prioritize processing `response.authority` over `response.answer`
- Verify the code is processing a `dns.rdtypes.ANY.SOA.SOA` before trying to extract `MNAME`
   - Continue the loop if not since the response *should* include an `MNAME` value somewhere

The error can be reproduced and the fix verified using the command line `./dnsrecon.py -t std -d <domain_name>` and any of the following domains:

```
fonts.googleapis.com
itunes.apple.com
sso.ns.agency
```

*NOTE*: The changes to 215-216 fix an additional exception that is caused in certain circumstances where certain responses are empty. The exception is raised because in this situation the value of `answers` in never set.  This can be reproduced with the domain name `juridico.fomenko.ru`.

```
[*] Trying juridico.fomenko.ru
[+] 0 Records Found
[*] Performing General Enumeration of Domain:fomenko.ru
[-] All nameservers failed to answer the DNSSEC query for fomenko.ru
Traceback (most recent call last):
  File "/home/UserName/git/dnsrecon/dnsrecon.py", line 1699, in <module>
    main()
  File "/home/UserName/git/dnsrecon/dnsrecon.py", line 1530, in main
    std_enum_records = general_enum(res, domain, xfr, goo, bing, spf_enum, do_whois, do_crt, zonewalk)
  File "/home/UserName/git/dnsrecon/dnsrecon.py", line 943, in general_enum
    found_soa_records = res.get_soa()
  File "/home/UserName/git/dnsrecon/lib/dnshelper.py", line 215, in get_soa
    for rdata in answers:
UnboundLocalError: local variable 'answers' referenced before assignment
```

### CODE REVIEW

File: lib/dnshelper.py [hunk 1]
```diff
@@ -207,33 +207,45 @@ def get_soa(self):
...
-                    if name.endswith('.'):
-                        soa_records.append(['SOA', name[:-1], ardata.address])
-                    else:
-                        soa_records.append(['SOA', name, ardata.address])
+                name = None
+                if isinstance(rdata, dns.rdtypes.ANY.SOA.SOA):
+                    name = rdata.mname.to_text()
+
+                if name:
+                    ipv6_answers = self._res.query(name, 'AAAA', tcp=tcp)
```
Comment (tsellers-r7) [inline lib/dnshelper.py]: Note the changes in this line and below are just changing `ipv4_answers` to `ipv6_answers` to reflect the `AAAA` query.

**Comment by tsellers-r7** [comment]
I just noticed that this PR also includes the same fix found in https://github.com/darkoperator/dnsrecon/pull/90

**Comment by tsellers-r7** [comment]
@darkoperator - I've addressed the Merge conflict, this should be a clean merge now.

**Comment by darkoperator** [comment]
Awesome!

**Pull request merged**